card_records.rb

```diff
@@ -0,0 +1,219 @@
+              )
+
+            card_params["document_url"] = "s3://bucket/#{AWS_BUCKET_PATH}/#{filename}"
+
+            file = download_pdf(card_params["location_url"])
+            upload_pdf(file, filename)
+          end
+
+          card_params["scrape_url"] = browser.url
+          card_params["well_files_url"] = well_files_url
+
+          NewMexicoCardRecord.new(card_params).save
+
+          browser.windows.last.close
+        end
+      end
+
+      close_browser
+    end
+
+    def filing_date(date)
+      date.strftime("%Y%m%d")
+    end
+
+    def filename(file_url: , document_type: , permit_id: , filing_date: )
+      "#{file_url.split("/").last[0...-4]}_#{document_type}_#{permit_id}_#{filing_date}.pdf"
+    end
+
+    def strpdate(date)
+      return "" unless date.present?
+
+      Date.strptime(date, "%m/%d/%Y")
+    end
+
+    def pagination
+      browser.element(css: "ul.pagination")
+    end
+
+    def count_of_pages
+      browser.element(css: "span#ctl00_ctl00__main_main_pager_lblPageCount").text.split(" ").last.to_i
+    end
+
+    def fill_form
+      current_year = Date.current.strftime("%Y")
+
+      browser.select(name: "ctl00$ctl00$_main$main$ddlPermitType").select("All")
+      browser.select(name: "ctl00$ctl00$_main$main$ddlPermitStatus").select("All")
+      browser.select(name: "ctl00$ctl00$_main$main$ddlStatusYear").select(current_year)
+    end
+
+    def table_params
+      browser.windows.last.use
+
+      browser.wait_until { |b| b.fieldsets.first.exists? }
+
+      browser.tables.first.rows.map do |row|
+        row.cells.map do |c|
+          c.attribute_value('textContent').squish
+        end
+      end.to_h.transform_keys do |key|
+        RENAME2.fetch(key, "")
+      end.select { |key, value| key.present? }
+    end
+
+    def row_params(row)
+      [
+        *headers.zip(cells_text_of_row(row))
+      ].to_h.transform_keys { |key| RENAME.fetch(key) }
+    end
+
+    def cells_text_of_row(row)
+      row.cells.map(&:text)
+    end
+
+    def location_url
+      browser.fieldsets.last.links.first.href
+    end
+
+    def well_files_url
+      button = browser.button(name: "ctl00$ctl00$_main$main$btnWellFile")
+
+      if button.exists?
+        button.onclick.match(/'(.*?)'/)[1]
+      else
+        ""
+      end
+    end
+
+    def headers
+      browser.table.headers.map(&:text)
+    end
+
+    def rows
+      browser.table.rows[1..-2]
+    end
+
+    def submit_form
+      browser.button(name: "ctl00$ctl00$_main$main$btnfilter").click
+    end
+
+    def uri
+      URI.join(HOST, PERMIT_PATH).to_s
+    end
+
+    def download_pdf(url)
+      open(url)
+    rescue => e
+      Rails.logger.info("Download failed - #{e}")
+      raise
+    end
+
+    def upload_pdf(file, filename)
+      Rails.logger.info("Uploading to S3 - #{filename}.pdf")
+      client.put_object(
+        body: file.read,
+        bucket: "bucket",
+        key: "#{AWS_BUCKET_PATH}#{filename}.pdf",
```

Comment:
the filename function already puts a "pdf" on the end right? I like that better than appending it here.